File: mtc-model/src/main/scala/org/drugis/mtc/yadas/YadasResults.scala
/*
 * This file is part of drugis.org MTC.
 * MTC is distributed from http://drugis.org/mtc.
 * Copyright (C) 2009-2010 Gert van Valkenhoef.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.drugis.mtc.yadas

import scala.collection.mutable.ArrayBuffer
import org.drugis.mtc.{Parameter,MCMCResults,
	MCMCResultsListener,MCMCResultsEvent}
import gov.lanl.yadas.MCMCParameter

abstract class ParameterWriter(val p: MCMCParameter, val i: Int) {
	def output() {
		write(p.getValue(i))
	}

	def write(x: Double): Unit
}

class Derivation(val pmtz: Map[Parameter, Int]) {
	assert(!pmtz.isEmpty)

	def calculate(results: MCMCResults, c: Int): Seq[Double] = {
		(0 until results.getNumberOfSamples).map(i => calculate(results, c, i))
	}
	def calculate(results: MCMCResults, c: Int, i: Int): Double = {
		pmtz.keySet.toList.map(p => pmtz(p) * results.getSample(results.findParameter(p), c, i)).reduceLeft((a, b) => a + b)
	}
}

class RandomEffectsVariance extends Parameter {
	override def getName = "var.d"
	override def toString = getName
}

class InconsistencyVariance extends Parameter {
	override def getName = "var.w"
	override def toString = getName
}

class YadasResults extends MCMCResults {
	private var results: List[List[ArrayBuffer[Double]]] = List()
	private var directParameters: List[Parameter] = List()
	private var derivedParameters: List[Parameter] = List()
	private var derivations: List[Derivation] = List()
	private var nChains: Int = 0
	private var reservedSamples: Int = 0
	private var availableSamples: Int = 0

	private class YadasParameterWriter(
		val paramIdx: Int, val chainIdx: Int, mp: MCMCParameter, i: Int)
	extends ParameterWriter(mp, i) {
		private var idx: Int = 0

		override def write(x: Double): Unit = {
			results(chainIdx)(paramIdx)(idx) = x
			idx += 1
		}
	}

	def setDirectParameters(p: List[Parameter]): Unit = {
		directParameters = p
		initResults()
	}

	def setDerivedParameters(p: List[(Parameter, Derivation)]): Unit = {
		derivedParameters = p.map(x => x._1)
		derivations = p.map(x => x._2)
	}

	def setNumberOfChains(n: Int) {
		nChains = n
		initResults()
	}

	def setNumberOfIterations(n: Int) {
		reservedSamples = n
		results = results.map(chain => chain.map(param => param.padTo(n, 0.0)))
	}

	def simulationFinished() {
		availableSamples = reservedSamples
		fireResultsChanged()
	}

	private def initResults(): Unit = {
		results = (0 until nChains).map(x => newChain).toList
	}

	private def newChain: List[ArrayBuffer[Double]] = {
		directParameters.map(x => new ArrayBuffer[Double](reservedSamples).padTo(reservedSamples, 0.0))
	}

	/**
	 * Writer to write samples from mp[i] to parameter p, chain c.
	 */
	def getParameterWriter(p: Parameter, c: Int, mp: MCMCParameter, i: Int)
	: ParameterWriter = new YadasParameterWriter(findParameter(p), c, mp, i)

	def getParameters: Array[Parameter] = directParameters.toArray
	def findParameter(p: Parameter): Int = 
		(directParameters ++ derivedParameters).findIndexOf(x => x == p)
	def getNumberOfChains: Int = nChains
	def getNumberOfSamples: Int = availableSamples
	def getSample(p: Int, c: Int, i: Int): Double = {
		assertBounds(p, c)
		if (p < directParameters.size) {
			results(c)(p)(i)
		} else {
			derivations(p - directParameters.size).calculate(this, c, i)
		}
	}
	def getSamples(p: Int, c: Int): Array[Double] = {
		assertBounds(p, c)
		if (p < directParameters.size) {
			results(c)(p).toArray
		} else {
			derivations(p - directParameters.size).calculate(this, c).toArray
		}
	}

	private def assertBounds(p: Int, c: Int) {
		if (c < 0 || c >= nChains) {
			throw new IndexOutOfBoundsException("Chain " + c + " is out of bounds (" + nChains + " chains)")
		}
		if (p < 0 || p >= directParameters.size + derivations.size) {
			throw new IndexOutOfBoundsException("Parameter " + p + " is out of bounds (" + directParameters.size + " + " + derivations.size + " parameters)")
		}
	}

	val listeners = new ArrayBuffer[MCMCResultsListener]()

	def addResultsListener(l: MCMCResultsListener) {
		listeners += l
	}

	def removeResultsListener(l: MCMCResultsListener) {
		listeners -= l
	}

	def clear() {
		results = null
		availableSamples = 0
	}

	private def fireResultsChanged() {
		for (l <- listeners) {
			l.resultsEvent(new MCMCResultsEvent(this))
		}
	}
}

File: mtc-mcmc/src/main/java/org/drugis/mtc/yadas/YadasResults.java
/*
 * This file is part of the GeMTC software for MTC model generation and
 * analysis. GeMTC is distributed from http://drugis.org/gemtc.
 * Copyright (C) 2009-2012 Gert van Valkenhoef.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.drugis.mtc.yadas;

import gov.lanl.yadas.MCMCParameter;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.apache.commons.lang.ArrayUtils;
import org.drugis.mtc.MCMCResults;
import org.drugis.mtc.MCMCResultsEvent;
import org.drugis.mtc.MCMCResultsListener;
import org.drugis.mtc.Parameter;

public class YadasResults implements MCMCResults {
	private Parameter[] d_directParameters;
	private Parameter[] d_derivedParameters;
	private Derivation[] d_derivations;
	private int d_nChains;
	private int d_availableSamples;
	private int d_reservedSamples;
	private List<List<double[]>> d_results;
	private List<MCMCResultsListener> d_listeners = new ArrayList<MCMCResultsListener>();
	
	private class YadasParameterWriter extends ParameterWriter {
		private final int d_pIdx;
		private final int d_cIdx;
		private int d_idx;

		/**
		 * @param pIdx Index of the parameter in the results.
		 * @param cIdx Chain index.
		 * @param mp MCMCParameter to monitor.
		 * @param i The component of the MCMCParameter to monitor.
		 */
		public YadasParameterWriter(int pIdx, int cIdx, MCMCParameter mp, int i) {
			super(mp, i);
			d_pIdx = pIdx;
			d_cIdx = cIdx;
			d_idx = 0;
		}

		@Override
		protected void write(double value) {
			d_results.get(d_cIdx).get(d_pIdx)[d_idx++] = value;
		}
	}
	
	public YadasResults() {
		initialize();
	}

	@Override
	public void clear() {
		d_availableSamples = 0;
		d_reservedSamples = 0;
		initResults();
	}
	
	private void initialize() { 
		d_directParameters = new Parameter[] {};
		d_derivedParameters = new Parameter[] {};
		d_derivations = new Derivation[] {};
		d_nChains = 0;
		clear();
	}

	/**
	 * Set the parameters for which results will be produced by the Yadas simulation.
	 */
	public void setDirectParameters(List<Parameter> parameters) {
		d_directParameters = new Parameter[parameters.size()];
		for (int i = 0; i < parameters.size(); ++i) {
			d_directParameters[i] = parameters.get(i);
		}
		initResults();
	}

	/**
	 * Set the derived parameters, which are to be calculated from the direct ones using the given Derivation.
	 */
	public void setDerivedParameters(Map<? extends Parameter, Derivation> derivations) {
		d_derivedParameters = new Parameter[derivations.size()];
		d_derivations = new Derivation[derivations.size()];
		int i = 0;
		for (Entry<? extends Parameter, Derivation> e : derivations.entrySet()) {
			d_derivedParameters[i] = e.getKey();
			d_derivations[i] = e.getValue();
			++i;
		}
	}

	/**
	 * Get the direct parameters.
	 * @see MCMCResults
	 */
	@Override
	public Parameter[] getParameters() {
		return d_directParameters;
	}

	@Override
	public int findParameter(Parameter p) {
		int idx = ArrayUtils.indexOf(d_directParameters, p);
		if (idx >= 0) {
			return idx;
		}
		idx = ArrayUtils.indexOf(d_derivedParameters, p);
		if (idx >= 0) {
			idx += d_directParameters.length;
		}
		return idx;
	}

	/**
	 * Create the ParameterWriter for the given Parameter.
	 * @param p Model parameter to write results for.
	 * @param cIdx Chain index.
	 * @param mp MCMCParameter to monitor results from.
	 * @param i Component of the MCMCParameter to monitor.
	 * @return A new ParameterWriter.
	 */
	public ParameterWriter getParameterWriter(Parameter p, int cIdx, MCMCParameter mp, int i) {
		return new YadasParameterWriter(findParameter(p), cIdx, mp, i);
	}

	@Override
	public double getSample(int p, int c, int i) {
		if (c < 0 || c >= d_nChains) {
			throw new IndexOutOfBoundsException("Chain " + c + " out of bounds: " +
					d_nChains + " chains available.");
		}
		final int nDirect = d_directParameters.length;
		final int nDerived = d_derivedParameters.length;
		if (p < 0 || p >= nDirect + nDerived) {
			throw new IndexOutOfBoundsException("Parameter " + p + " out of bounds: " + nDirect +
					" + " + nDerived + " parameters available.");
		}
		if (i < 0 || i >= d_availableSamples) {
			throw new IndexOutOfBoundsException("Iteration " + i + " out of bounds: " +
					d_availableSamples + " iterations available.");
		}
		return p < nDirect ? d_results.get(c).get(p)[i] : d_derivations[p - nDirect].calculate(this, c, i);
	}


	public void setNumberOfChains(int nChains) {
		d_nChains = nChains;
		initResults();
	}

	private void initResults() {
		d_results = new ArrayList<List<double[]>>(d_nChains);
		for (int i = 0; i < d_nChains; ++i) {
			List<double[]> list = new ArrayList<double[]>(d_directParameters.length);
			for (int j = 0; j < d_directParameters.length; ++j) {
				list.add(new double[d_reservedSamples]);
			}
			d_results.add(list);
		}
	}

	@Override
	public int getNumberOfChains() {
		return d_nChains;
	}
	
	public void setNumberOfIterations(int iter) {
		d_reservedSamples = iter;
		for (List<double[]> chain : d_results) {
			for (int i = 0; i < chain.size(); ++i) {
				double[] newResults = new double[iter];
				System.arraycopy(chain.get(i), 0, newResults, 0, chain.get(i).length);
				chain.set(i, newResults);
			}
		}
	}
	
	public int getNumberOfIterations() { 
		return d_reservedSamples;
	}

	@Override
	public int getNumberOfSamples() {
		return d_availableSamples;
	}
	
	public void simulationFinished() {
		d_availableSamples = d_reservedSamples;
		MCMCResultsEvent event = new MCMCResultsEvent(this);
		List<MCMCResultsListener> listeners = new ArrayList<MCMCResultsListener>(d_listeners);
		for (MCMCResultsListener l : listeners) {
			l.resultsEvent(event);
		}
	}
	

	
	@Override
	public void addResultsListener(MCMCResultsListener l) {
		d_listeners.add(l);
	}
	
	@Override
	public void removeResultsListener(MCMCResultsListener l) {
		d_listeners.remove(l);
	}
}
